File: test/status.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import {hasChangesOrDivergence, parseStatus} from "../src/status.ts";

test("clean branch-only status is unchanged", () => {
  const parsed = parseStatus("## main...origin/main\n");
  assert.equal(parsed.changed, false);
  assert.equal(hasChangesOrDivergence(parsed.raw), false);
  assert.equal(parsed.branch?.name, "main");
  assert.equal(parsed.branch?.upstream, "origin/main");
});

test("ahead and behind branch status counts as changed", () => {
  const parsed = parseStatus("## main...origin/main [ahead 1, behind 2]\n");
  assert.equal(parsed.changed, true);
  assert.equal(parsed.branch?.ahead, 1);
  assert.equal(parsed.branch?.behind, 2);
});

test("gone upstream counts as changed", () => {
  const parsed = parseStatus("## feature...origin/feature [gone]\n");
  assert.equal(parsed.changed, true);
  assert.equal(parsed.branch?.gone, true);
});

test("parses modified file", () => {
  const parsed = parseStatus("## main\n M src/index.ts\n");
  assert.equal(parsed.changed, true);
  assert.equal(parsed.items[0]?.category, "modified");
});

test("parses staged file", () => {
  const parsed = parseStatus("## main\nA  README.md\n");
  assert.equal(parsed.items[0]?.category, "staged");
});

test("parses untracked file", () => {
  const parsed = parseStatus("## main\n?? README.md\n");
  assert.equal(parsed.items[0]?.category, "untracked");
});

test("parses deleted file", () => {
  const parsed = parseStatus("## main\n D old.ts\n");
  assert.equal(parsed.items[0]?.category, "deleted");
});

test("parses renamed file", () => {
  const parsed = parseStatus("## main\nR  old.ts -> new.ts\n");
  assert.equal(parsed.items[0]?.category, "renamed");
});

test("parses conflict", () => {
  const parsed = parseStatus("## main\nUU conflicted.ts\n");
  assert.equal(parsed.items[0]?.category, "conflict");
});

File: internal/status/status_test.go
package status

import "testing"

func TestCleanBranchOnlyStatusIsUnchanged(t *testing.T) {
	parsed := Parse("## main...origin/main\n")
	if parsed.Changed || HasChangesOrDivergence(parsed.Raw) {
		t.Fatal("expected clean branch-only status to be unchanged")
	}
	if parsed.Branch == nil || parsed.Branch.Name != "main" || parsed.Branch.Upstream != "origin/main" {
		t.Fatalf("unexpected branch: %+v", parsed.Branch)
	}
}

func TestAheadAndBehindCountsAsChanged(t *testing.T) {
	parsed := Parse("## main...origin/main [ahead 1, behind 2]\n")
	if !parsed.Changed {
		t.Fatal("expected ahead/behind to count as changed")
	}
	if parsed.Branch == nil || parsed.Branch.Ahead != 1 || parsed.Branch.Behind != 2 {
		t.Fatalf("unexpected branch: %+v", parsed.Branch)
	}
}

func TestGoneUpstreamCountsAsChanged(t *testing.T) {
	parsed := Parse("## feature...origin/feature [gone]\n")
	if !parsed.Changed || parsed.Branch == nil || !parsed.Branch.Gone {
		t.Fatalf("expected gone upstream: %+v", parsed.Branch)
	}
}

func TestParsesStatusCategories(t *testing.T) {
	tests := []struct {
		name string
		raw  string
		want Category
	}{
		{name: "modified", raw: "## main\n M src/index.ts\n", want: Modified},
		{name: "staged", raw: "## main\nA  README.md\n", want: Staged},
		{name: "untracked", raw: "## main\n?? README.md\n", want: Untracked},
		{name: "deleted", raw: "## main\n D old.ts\n", want: Deleted},
		{name: "renamed", raw: "## main\nR  old.ts -> new.ts\n", want: Renamed},
		{name: "conflict", raw: "## main\nUU conflicted.ts\n", want: Conflict},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			parsed := Parse(test.raw)
			if len(parsed.Items) != 1 {
				t.Fatalf("expected one item, got %d", len(parsed.Items))
			}
			if parsed.Items[0].Category != test.want {
				t.Fatalf("expected %s, got %s", test.want, parsed.Items[0].Category)
			}
		})
	}
}
